### PR TITLE
use sigaction instead of signal

### LINKED ISSUE #16: Use sigaction without SA_RESTART instead?
Currently it uses the `signal` function which translates to `sigaction` with the SA_RESTART flag. Most blocking syscalls aren't interrupted by it, e.g. reading from a file descriptor will continue to wait for data after the signal. Setting a `running` flag won't work unless that read returns and the process can check the `running` flag.

I encountered this when I was reading inotify events. <kbd>Ctrl-C</kbd> didn't stop it until the next event arrived. I'm on Arch Linux x86_64 if it matters.

The change won't be backwards compatible, so perhaps add another feature to opt in?

### PR DESCRIPTION
**Pull request by lilydjwg**
closes #16

### CODE REVIEW

**Comment by henninglive** [comment]
I don't like disabling SA_RESTART by default.

**Comment by henninglive** [comment]
What should happen if there is another handler, this will overwrite it without an error, is this what we want? We could try restoring it and return an error or if it’s an SA_SIGINFO handler, we could try chaining it after our own handler.

**Comment by lilydjwg** [comment]
I've put that behind a feature.

Maybe the caller should take aware of that. And I find that it's not very reliable in my case. I have to sleep a short period after receiving EINTR, because the `running` flag may not have flipped at the time.

```
    while running.load(Ordering::SeqCst) {
      if let Err(e) = notify.read_events_blocking(&mut buf) {
        if e.kind() == io::ErrorKind::Interrupted {
          std::thread::sleep(std::time::Duration::from_millis(10));
          continue;
        } else {
          return Err(e)?;
        }
      };
      // do things
    }
```

Maybe it's better to use tokio and signalfd for more sophisticated cases.

**Comment by henninglive** [comment]
The sleep is necessary since our handler runs on its own thread, there is going to be some delay before the OS schedules it to run. If you want the running flag to be updated immediately when your system calls gets interrupted you need to set the running flag in the os_handler(). Running our handler in its own thread is unfortunate, but we cannot safely allow arbitrary rust code to run in the os_handler() since most operations are not async signal safe.

File: src/lib.rs [hunk 1]
```diff
@@ -191,6 +204,22 @@ pub fn set_handler<F>(user_handler: F) -> Result<(), Error>
...
+            unsafe {
+                let mut sigset = ::std::mem::uninitialized::
+                    <libc::sigset_t>();
+                libc::sigemptyset(&mut sigset);
+                libc::sigaddset(&mut sigset, libc::SIGINT);
+                #[cfg(feature = "termination")]
+                libc::sigaddset(&mut sigset, libc::SIGTERM);
+                libc::pthread_sigmask(
+                    libc::SIG_BLOCK, &sigset, ::std::ptr::null_mut());
+            }
```
Comment (henninglive) [inline src/lib.rs]: Why do we need to do this? This will block the signals on this thread only, right?

Comment (lilydjwg) [inline src/lib.rs]: Right. In a multithreaded process, the signal may be directed to an arbitrary thread, and only that thread will be interrupted. When we want `EINTR`, we definitely want it to be directed to some work thread, not this one.

**Comment by henninglive** [comment]
After thinking this over, I am not sure this is something we should try to support. As you point out, using EINTR in this way requires that the user take special care to make sure that signals are delivered to the right thread. Either by masking it them to the right thread or by using pthread_kill() signal a specific thread. If a user wants to do this, they should probably handle this themselves instead of using this crate.

**Comment by lilydjwg** [comment]
I'm starting to think, how useful is it if one can't interrupt a blocking call on Ctrl-C?

On crates.io there are five crates depending on ctrlc. Three of them (gauc, panty and amber) kill things, remove files and then exit, one (pcp) is computing things so it doesn't get blocked on syscalls, one (connectr) checks back every 100ms.

With the first kind, it's fine as long as the cleanup action is simple. With the second, it wastes CPU cycles but it works. With the third, it works too but it may use too many syscalls than necessary, which makes strace logs harder to read. Also it makes the application a little slow to response.

I guess it would be better if what ctrlc does is only to flip a AtomicBool instead, so users don't need an extra thread and pair of pipes (which is a little annoying when I lsof the process).

**Comment by henninglive** [comment]
I agree, dropping this reduces usefulness, but I feel like including this creates an unsafe and leaky interface. This feature is very dependent on threading setup of program, depending on how many threads are in use, the user might need to call external libc functions themself to make sure the signals are delivered to the right thread.

Generally, if one needs cancel IO, one should probably be using [tokio](https://github.com/tokio-rs/tokio) and [tokio-signal](https://github.com/alexcrichton/tokio-signal).

**Comment by henninglive** [comment]
Regarding how dependent crates are using this carte. We could abstract over some of the common usage patterns. For example, we could add a CtrlGuard with a block method that blocks until a signal is received. This would be implemented without the extra thread. We could also add CtrlAtomic(Bool/Int) or something, that implement a flag that gets updated when a signal is received. This would be implemented without the extra thread, pips or semaphores.

**Comment by lilydjwg** [comment]
I give up. Most functions in std ignore `EINTR`, to make this mostly useless.

**Pull request closed without merging**